### DISCUSSION TITLE
Any reason to install when using pnpm?

### DISCUSSION THREAD

**Discussion by IanVS** [biomejs/biome#1986] [General]
The getting started guide says to use `pnpm add --save-dev --save-exact @biomejs/biome`, which will install the project in node_modules, but then all of the instructions for running biome use `pnpm dlx`, which will download and execute the latest version each time it is run, which defeats the purpose of saving an exact version in node_modules.  Instead, a version should be specified like `pnpm dlx @biomejs/biome@1.5.3 check .`.

Side note: why is `dlx` required to run biome?

**Comment by jasikpark** [comment]
The answer is `pnpm exec biome [...args]` or just `biome` when in a `scripts`

https://discord.com/channels/1132231889290285117/1132602515616174080/1215058401415925900

**Reply by IanVS** [reply]
Cool, I think the documentation should be updated to reflect that, then.  @ematipico would you accept a PR updating the docs from `pnpm dlx @biomejs/biome …` to `pnpm biome …`?

**Reply by ematipico** [reply]
I suppose you're right, but it's unsettling that `pnpm dlx` doesn't behave the same as `npx`.

Plus, unfortunately, the binary name isn't the same as the package name, so I'm concerned this could cause more confusion to users. I've seen many users writing `npx biome format`, which is wrong.

We could **add** `pnpm exec` and add a comment that explains said command.

**Reply by arendjr** [reply]
I don’t use pnpm, but I see the Yarn instructions have the same issue.

> I suppose you're right, but it's unsettling that `pnpm dlx` doesn't behave the same as `npx`.

Yarn and pnpm seem to be aligned here. The unsettling part is that `npx` works the way it does. It conditions users to use `npx` for running local scripts, but the moment they make a typo they’ll be running code from possibly typo-squatted packages on their machine.

If the owner of the `biome` package ever becomes malicious, anyone running `npx biome` may have a serious problem (and by extension, us).

I think it’s safer to use `yarn biome`/`pnpm biome` in their respective instructions ~~and use `npm run biome` for the npm instructions~~. **Edit:** Unfortunately, `npm run` doesn't work for installed executables and confusingly, `npx biome` *does* work after Biome is installed, although it risks running the wrong package as long as Biome isn't installed. So I guess we better stick with `npx @biomejs/biome` there.

**Reply by hilja** [reply]
`npx` is two commands rolled into one, which can be confusing if you only want to run the other. I guess the pnpm philosophy is to have less side-effects and it splits it to `pnpm dlx` and `pnpm exec`.

After `pnpm add -D biome`, you can just do `pnpm biome`, the exec is implied. I've gotten used to running most clis like that.

Tangent: in the [v9](https://github.com/pnpm/pnpm/releases/tag/v9.0.0-alpha.7) they will finally set `enable-pre-post-scripts` to true by default, I guess it's good to allow some side-effects :)

**Reply by ematipico** [reply]
I **think** the majority of the users don't know the distinction between package name and binary name, plus we are in an awkward position because in 99% of the cases, the package name matches with the name of the binary, which isn't our case.

Plus, `bunx` works as `npx`.

I see two options:
- get rid of the installation step
- expand the pnpm/yarn commands to show alternatives: `pnpm biome`/`pnpm exec` and ditch the `dlx`

**Reply by arendjr** [reply]
I vote for option 2. That way we just document those tools the way they are supposed to be used. Using `dlx` just doesn’t make a lot of sense here since it doesn’t behave according to the workflow we advise for the other tools. Happy to make the PR for this.

**Reply by ematipico** [reply]
Thank you!

**Reply by arendjr** [reply]
https://github.com/biomejs/biome/pull/2101

**Comment by arendjr** [comment]
The documentation was updated in https://github.com/biomejs/biome/pull/2101.
